src/server/editorServices.ts

```diff
@@ -464,6 +464,11 @@ namespace ts.server {
         return copiedList;
     }
 
+    export function forEachProject<T>(projects: Project[], action: (project: Project) => T[], comparer?: (a: T, b: T) => number, areEqual?: (a: T, b: T) => boolean) {

```

Comment:
This function does a *lot* and doesn't reflect it in its name.

You are doing a concat-map, then sorting the result, and then deduplicating? That doesn't sound at all like what a `forEach` is going to do - not to mention that the `forEach` in `core.ts` already has really wonky but different semantics.